Detect & Respond sections via @thatsjet (#840)
OWASP/CheatSheetSeries: cheatsheets_draft/Secrets_Management_CheatSheet.md

- Key handling and cryptographic implementations are taken care of by Encryption Service, not by developers
- The provider could add more services to interact with the sensitive data

> ## 8 Detection

Reviewer: Could you update the index up top to match this title?

Reviewer 2: Will do in a next iteration :)

> There are many approaches to secrets detection and some very useful open source projects to help with this. Of note, the [Yelp Detect Secrets](https://github.com/Yelp/detect-secrets) project is mature and has signature matching for around 20 secrets. For more information other tools to help you in the detection space check out the [Secrets Detection](https://github.com/topics/secrets-detection) topic on GitHub.

Reviewer: Small tweak for conciseness, add missing word
```suggestion
There are many approaches to secrets detection and some very useful open source projects to help with this. The [Yelp Detect Secrets](https://github.com/Yelp/detect-secrets) project is mature and has signature matching for around 20 secrets. For more information on other tools to help you in the detection space, check out the [Secrets Detection](https://github.com/topics/secrets-detection) topic on GitHub.
```

### 8.1 General detection approaches

> Shift-left and DevSecOps principals apply to secrets detection as well. These general approaches below are aimed at considering secrets earlier and evolving the practice over time.

Reviewer: Fix typo, rewrite to active.
```suggestion
Shift-left and DevSecOps principles apply to secrets detection as well. These general approaches below aim to consider secrets earlier and evolve the practice over time.
```

> - Create standard test secrets and use them universally across the organization. This allows for the reduction of false-positives by only needing to track a single test secret for each secret type.

Reviewer: Hyphen not necessary, conciseness
```suggestion
- Create standard test secrets and use them universally across the organization. This allows for reducing false positives by only needing to track a single test secret for each secret type.
```

- Consider enabling secrets detection at the developer level to avoid checking secrets into code before commit/PR either in the IDE, as part of test-driven development, or via pre-commit hook.
- Make secrets detection part of the threat model. Consider secrets as part of the attack surface during threat modeling exercises.
- Evaluate detection utilities and related signatures often to ensure they meet expectations.
- Consider having more than one detection utility and correlating/de-duping results to identify potential areas of detection weakness.

> - Also consider exploring a balance between entropy and ease of detection. Secrets with consistent formats are easier to detect with lower false-positive rates, but you also don't want to miss a human-created password simply because it doesn't match your detection rules.

Reviewer: Conciseness
```suggestion
- Explore a balance between entropy and ease of detection. Secrets with consistent formats are easier to detect with lower false-positive rates, but you also don't want to miss a human-created password simply because it doesn't match your detection rules.
```

### 8.2 Types of secrets to be detected

> Many types of secrets exists and signatures for each should be considered to ensure accurate detection for all. Among the more common types are:

Reviewer: Fix typo, passive to active
```suggestion
Many types of secrets exist, and you should consider signatures for each to ensure accurate detection for all. Among the more common types are:
```

- High availability secrets (Tokens that are difficult to rotate)
- Application configuration files
- Connection strings
- API keys
- Credentials
- Passwords
- 2FA keys
- Private keys (like ssh keys)
- Session tokens

> - Platform specific secret types (e.g.; Amazon Web Services, Google Cloud)

Reviewer: Add hyphen
```suggestion
- Platform-specific secret types (e.g., Amazon Web Services, Google Cloud)
```

For more fun learning about secrets and practice rooting them out check out the [Wrong Secrets](https://owasp.org/www-project-wrongsecrets/) project.

### 8.3 Detection lifecycle

Secrets are like any other authorization token. They should:

- Exist only for as long as necessary (rotate often)
- Have a method for automatic rotation
- Only be visible to those who need them (least privilege)
- Be revokable (including the logging of attempt to use a revoked secret)
- Never be logged (must implement either an encryption or masking approach in place to avoid logging plaintext secrets)

> Detections should exist during each of these stages of the secret lifecycle to ensure secrets were not introduced but missed in a prior step.

Reviewer: I find this sentence a bit unclear, maybe rewrite or shorten it?
```suggestion
Create detection rules for each of the stages of the secret lifecycle.
```

### 8.4 Documentation for how to detect secrets

> Ensure documentation is created and updated regularly to inform the developer community on procedures and systems available at your company and what types of secrets management is expected, how to test for secrets, and what to do in event of secrets detected.

Reviewer: Rewrite to active
```suggestion
Create documentation and update it regularly to inform the developer community on procedures and systems available at your company and what types of secrets management you expect, how to test for secrets, and what to do in event of detected secrets.
```

Documentation should:

- Exist and be updated often, especially in response to an incident
- Include the following information:
    - Who has access to the secret
    - How it gets rotated
    - Any upstream or downstream dependencies that could potentially be broken during secret rotation
    - Who is the point of contact during an incident
    - Security impact of exposure

- Identify when secrets may be handled differently depending on the threat risk, data classification, etc.

> ## 9 Incident Response

Reviewer: Could you edit the index up top to match this title?

Reviewer 2: Will do in final pr

> Quick response in the event of a secret being exposed is perhaps one of the most important considerations for secrets management.

Reviewer: ```suggestion
Quick response in the event of a secret exposure is perhaps one of the most critical considerations for secrets management.
```

### 9.1 Documentation

> The primary consideration for responding to the exposure of a secret is to ensure that everyone in the chain of custody is aware of and understands how to respond in the event of exposure. This includes application creators (every member of a development team), information security, and technology leadership.

Reviewer: Somewhat wordy, concise rewrite?
```suggestion
Incident response in the event of secret exposure should ensure that everyone in the chain of custody is aware and understands how to respond. This includes application creators (every member of a development team), information security, and technology leadership.
```

Documentation must include:

- How to test for secrets and secrets handling, especially during business continuity reviews.
- Whom to alert when a secret is detected.
- Steps to take for containment
- Information to log during the event

### 9.2 Remediation

The primary goal of incident response is rapid response and containment. 

Containment should follow these procedures:

> 1. Revocation: Keys that were exposed should ensure immediate revocation. They secret must be able to be de-authorized quickly and systems must be in place to identify the revocation status of a secret.

Reviewer: Typo, conciseness
```suggestion
1. Revocation: Keys that were exposed should ensure immediate revocation. The secret must be able to be de-authorized quickly, and systems must be in place to identify the revocation status.
```

1. Rotation: A new secret must be able to be quickly created and implemented, preferably via an automated process to ensure repeatability, low rate of implementation error, and least-privilege (not directly human-readable).

> 1. Deletion: Secrets revoked/rotated must be removed from the exposed system immediately, this includes secrets discovered in code or logs. Secrets in code should have commit history for the exposure squashed to before the introduction of the secret, and logs must have a process for removing the secret while maintaining log integrity.

Reviewer: conciseness
```suggestion
1. Deletion: Secrets revoked/rotated must be removed from the exposed system immediately, including secrets discovered in code or logs. Secrets in code should have commit history for the exposure squashed to before the introduction of the secret, and logs must have a process for removing the secret while maintaining log integrity.
```

> 1. Logging: Incident response teams must have access to information about the lifecycle of a secret to aid in containment and remediation including (see more in logging section below):

Reviewer: Section is short enough, could remove
```suggestion
1. Logging: Incident response teams must have access to information about the lifecycle of a secret to aid in containment and remediation, including:
```

    - Who had access?
    - When did they use it?
    - When was it previously rotated?

### 9.3 Logging

Additional considerations for logging of secrets usage should include:

- Logging for incident response should be to a single location accessible by IR teams

> - Fidelity of logging information should be ensured during purple team exercises such as:

Reviewer: Rewrite to active
```suggestion
- Ensure fidelity of logging information during purple team exercises such as:
```

    - What should have been logged?
    - What was actually logged?
    - Do we have adequate alerts in place to ensure this?

> Consider the use of a standardized logging format and vocabulary such as the [Application Logging Vocabulary Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Logging_Vocabulary_Cheat_Sheet.html) to ensure that all necessary information is logged.

Reviewer: conciseness
```suggestion
Consider using a standardized logging format and vocabulary such as the [Application Logging Vocabulary Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Logging_Vocabulary_Cheat_Sheet.html) to ensure that all necessary information is logged.
```

- Key handling and cryptographic implementations are taken care of by Encryption Service, not by developers
- The provider could add more services to interact with the sensitive data

## 8 Detection

There are many approaches to secrets detection and some very useful open source projects to help with this. Of note, the [Yelp Detect Secrets](https://github.com/Yelp/detect-secrets) project is mature and has signature matching for around 20 secrets.

### 8.1 General detection approaches

- Create standard test secrets and use them universally across the organization. This allows for the reduction of false-positives by only needing to track a single test secret for each secret type.
- Consider enabling secrets detection at the developer level to avoid checking secrets into code before commit/PR.
- Make secrets detection part of the threat model. Consider secrets as part of the attack surface during threat modeling exercises.
- Evaluate detection utilities and related signatures often to ensure they meet expectations.

> - Consider having more than one detection utility and correlating/de-duping results to identify potential areas of detection weakness.

Reviewer 2: good to note: man-made secrets (e.g. random passwordS) are often harder to detect than pre-defined API-key formats for which easy rules can be created. Maybe recommend finding a trade-off between entropy and the ease of detecting it?

### 8.2 Types of secrets to be detected

Many types of secrets exists and signatures for each should be considered to ensure accurate detection for all. Among the more common types are:

- High availability secrets (Tokens that are difficult to rotate)
- Application configuration files
- Connection strings
- Credentials
- Platform specific secret types (e.g.; Amazon Web Services, Google Cloud)

### 8.3 Detection lifecycle

Secrets are like any other authorization token. They should:

- Exist only for as long as necessary (rotate often)
- Have a method for automatic rotation
- Only be visible to those who need them (least privilege)
- Be revokable (including the logging of attempt to use a revoked secret)
- Never be logged

> Detections should exist during each of these stages of the secret lifecycle to ensure secrets were not introduced but missed in a prior step.

Reviewer 2: detection duing logging will mean that double-encoding and/or other things might need to be in palce as well?

Reviewer 2: and/or encryption
